### ISSUE TITLE
OAUTH_REDIRECT_AT_SIGN_IN results in 404 when using SAML

### ISSUE THREAD

**Issue by SISheogorath** [mastodon/mastodon#15959]
### Expected behaviour

When using `OAUTH_REDIRECT_AT_SIGN_IN` is set to true, while only a SAML login is configured, until version 3.3.0 a redirect to `<mastodon url>/auth/auth/saml` would take place when clicking on the "Login" button. This would automatically redirect the IdP and authenticate the user.

### Actual behaviour

Since version 3.3.0 when `OAUTH_REDIRECT_AT_SIGN_IN` is set to `true`, while only SAML login is configured, the login button will redirect to `<mastodon url>/auth/auth/saml`, but instead of redirecting to the IdP it show the error below on an otherwise white, unstyled page:

`Not found. Authentication passthru.`

In the Mastodon error log the following error appears:

```
[ca1a3458-4d23-46c8-bd86-31bb90593cba] method=GET path=/auth/sign_in format=html controller=Auth::SessionsController action=new status=302 duration=3.99 view=0.00 db=0.00 location=<mastodon url>/auth/auth/saml
[e94d59fd-3078-4ad2-9868-fc65f67259ca] method=GET path=/auth/auth/saml format=html controller=Auth::OmniauthCallbacksController action=passthru status=404 duration=2.77 view=0.59 db=0.00
[a20e0b37-38b6-4cef-b4c3-fc14bf970365] method=GET path=/auth/saml format=html controller=ApplicationController action=raise_not_found status=404 duration=31.76 view=26.86 db=0.00
```

### Steps to reproduce the problem

1. Configure a basic SAML login as shown in the `env.nanobox`:
https://github.com/tootsuite/mastodon/blob/a650a1157d5b71bc5718759643fe5bb572b16a8b/.env.nanobox#L232-L252

2. Set `OAUTH_REDIRECT_AT_SIGN_IN` to `true`

3. Click on login in the webui.

To prove it works otherwise, set `OAUTH_REDIRECT_AT_SIGN_IN` to `false` and try the same thing.

### Specifications

[v3.3.0](https://github.com/tootsuite/mastodon/tree/v3.3.0) in form of the [official docker image](https://hub.docker.com/r/tootsuite/mastodon)

**Comment by SISheogorath** [comment]
I want to note, that this issue still persists in 3.4.1.

**Comment by SISheogorath** [comment]
I'm not sure why it changed, but with some further investigation I figured out that when adjusting the following line to do a post request, it works again:

https://github.com/mastodon/mastodon/blob/7c27ccb8d8fda060f3c5b234f7c81fc29b5fc915/app/views/layouts/public.html.haml#L25

So basically the line above has to function like the line below:

https://github.com/mastodon/mastodon/blob/c3aef491d66aec743a3a53e934a494f653745b61/app/views/auth/sessions/new.html.haml#L25

And if I'm not mistaken, this commit introduced the breakage (intentional, since the goal was to force CSRF): https://github.com/mastodon/mastodon/commit/13b07b88f1aa79c31291473362ac77b31602c374

Given that the [introduced dependency, `omniauth-rails_csrf_protection`,](https://github.com/cookpad/omniauth-rails_csrf_protection/tree/v0.1.2) explicitly states:

> Disable access to the OAuth request phase using HTTP GET method.

It seems that adding the post request there, would be the right way to go, I just don't know ruby well enough to build that. 😅 Help would be very welcome, now that things are fully diagnosed.

**Comment by bobbyd0g** [comment]
I am also experiencing this issue and currently working around it with proxy redirects as best I can, would love to see it resolved as well as properly integrated into the upcoming OIDC support in #16221

**Comment by ClearlyClaire** [comment]
I'm not too familiar with neither the OAuth flow nor with OmniAuth/SAML, but my understanding of the issue is:
- #6540 added a `OAUTH_REDIRECT_AT_SIGN_IN` environment variable which is meant to redirect to an external authentication provider instead of presenting the usual log-in form whenever going to `/auth/sign_in`
- this `OAUTH_REDIRECT_AT_SIGN_IN` variable does not change anything to the log-in form (introduced in #10232) in the `/about` page, which was created later with no code specific to external auth
- this `OAUTH_REDIRECT_AT_SIGN_IN` has actually little to do with OAuth, unlike what its name suggests, as it applies to `/auth/sign_in`, regardless of how that was accessed, and OmniAuth doesn't necessarily use OAuth
- since #15228, accessing the OmniAuth routes is only possible through POST requests; this means the `OAUTH_REDIRECT_AT_SIGN_IN` is effectively broken, as the log-in form has to be accessed through `GET` and the redirect will therefore be redirected using `GET`
- changing the links to `/auth/sign_in` to do a `POST` request as suggested will not work, because a `POST` to that URL means something entirely different: submitting your credentials to actually log in
- furthermore, changing links to `/auth/sign_in` to use `POST` requests would not fix applications using OAuth either

Therefore, I think we should drop `OAUTH_REDIRECT_AT_SIGN_IN`, I don't think we can transparently redirect to an OmniAuth provider anymore (as it requires a POST, to avoid cross-site request forgery attacks that were discovered at the time), and the current behavior is broken.

**Comment by ClearlyClaire** [comment]
@SISheogorath as I wrote earlier:
- `POST`ing to `/auth/sign_in` does not trigger a redirect, it goes through another code path that is for handling submitting the actual login form, so to work it would require more changes; and even more changes to properly work with instances that do not have `OAUTH_REDIRECT_AT_SIGN_IN=true`
- this would cover links in the web pages itself, but not the OAuth flow, which is currently broken because it immediately redirects to `/auth/sign_in`

My plan is to later introduce something like `OMNIAUTH_ONLY=true` that would:
- disable sign-up
- change log-in forms to only display OmniAuth provider buttons
- change “Sign in” link generation to directly link to the OmniAuth provider if only one is configured (so it'd link directly to `/auth/saml` instead of `/auth/sign_in`

**Comment by SISheogorath** [comment]
Would this happen somewhere in the same release? Or is that a planned icebox feature for somewhere within the next decade maybe?

I mean, currently things are broken for almost a year now, so it probably doesn't make that much of a difference, as everyone got used to the extra clicks, it's still a feature that not just me is missing already.

**Comment by ClearlyClaire** [comment]
This is something I plan on working on soon (that is, within days), but I don't know how and when things will be merged and released. Anyway, I figure an extra click in the log-in flow is better than a completely broken one for the time being.

**Comment by ClearlyClaire** [comment]
@SISheogorath implemented in #17288 which replaces `OAUTH_REDIRECT_AT_SIGN_IN=true` with `OMNIAUTH_ONLY=true`. This does not automatically redirect anymore, but log-in links are replaced to directly point at the OmniAuth endpoint (using POST), and the regular log-in forms in `/about` and `/auth/sign_in` are replaced with the external provider buttons. `/auth/sign_in` is only accessed when being automatically redirected from a page requiring log-in, such as the OAuth flow if you aren't logged in already.

**Comment by InoryS** [comment]
Not sure if it's the right place
I'm trying to use OIDC with keycloak as Idp

I configured mastodon with following settings:


```
#OIDC
OIDC_ENABLED=true
OIDC_DISPLAY_NAME='name'
OIDC_ISSUER=https://keycloak/realms/master
OIDC_DISCOVERY=true
OIDC_SCOPE=openid,profile,email
OIDC_UID_FIELD=preferred_username
OIDC_CLIENT_ID=
OIDC_CLIENT_SECRET=
OIDC_REDIRECT_URI=https://mastodon/auth/auth/openid_connect/callback
OIDC_SECURITY_ASSUME_EMAIL_IS_VERIFIED=true
# uncomment to only use OIDC for login / registration buttons
OMNIAUTH_ONLY=false
```

But when I click login with 'name' nothing happens, I can see in the browser console that https://mastodon/auth/auth/openid_connect is being requested, this request never succeeds.

If I go directly access to https://mastodon/auth/auth/openid_connect I would get `"Not found. Authentication passthru."` or `"504 We're sorry, but something went wrong on our end."`

I can confirm that https://keycloak/realms/master/.well-known/openid-configuration is configured correctly.



The log of mastodon-web 
```
method=GET path=/auth/auth/openid_connect format=html controller=Auth::OmniauthCallbacksController action=passthru status=404 duration=2.86 view=1.41 db=0.00
method=GET path=/auth/auth/openid_connect format=html controller=Auth::OmniauthCallbacksController action=passthru status=404 duration=2.81 view=1.34 db=0.00
```
